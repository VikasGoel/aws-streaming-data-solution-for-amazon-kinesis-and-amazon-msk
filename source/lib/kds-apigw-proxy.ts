/*********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from '@aws-cdk/core';
import * as apigw from '@aws-cdk/aws-apigateway';
import * as iam from '@aws-cdk/aws-iam';
import * as cwlogs from '@aws-cdk/aws-logs';
import * as kinesis from '@aws-cdk/aws-kinesis';

export interface ProxyApiProps {
    readonly stream: kinesis.IStream;
    readonly accessLogsRetention: cwlogs.RetentionDays;
    readonly throttlingRateLimit: number;
    readonly throttlingBurstLimit: number;
}

interface IntegrationProps {
    readonly action: string;
    readonly path: string;
    readonly requestTemplate: string;
    readonly role: iam.Role;
    readonly model: apigw.IModel;
    readonly validator: apigw.IRequestValidator;
}

export class ProxyApi extends cdk.Construct {
    public readonly Api: apigw.RestApi;

    constructor(scope: cdk.Construct, id: string, props: ProxyApiProps) {
        super(scope, id);

        if (!cdk.Token.isUnresolved(props.throttlingRateLimit) && props.throttlingRateLimit <= 0) {
            throw new Error('throttlingRateLimit must be a positive number');
        }

        if (!cdk.Token.isUnresolved(props.throttlingBurstLimit) && props.throttlingBurstLimit < 0) {
            throw new Error('throttlingBurstLimit must be a non-negative number');
        }

        const logGroup = new cwlogs.LogGroup(this, 'AccessLogs', {
            retention: props.accessLogsRetention,
            removalPolicy: cdk.RemovalPolicy.RETAIN
        });

        const apiGatewayRole = new iam.Role(this, 'ApiGatewayRole', {
            assumedBy: new iam.ServicePrincipal('apigateway.amazonaws.com')
        });

        props.stream.grantWrite(apiGatewayRole);

        this.Api = new apigw.RestApi(this, 'Api', {
            restApiName: `${cdk.Aws.STACK_NAME}-kinesis-proxy`,
            cloudWatchRole: true,
            deployOptions: {
                methodOptions: {
                    '/*/*': {
                        throttlingRateLimit: props.throttlingRateLimit,
                        throttlingBurstLimit: props.throttlingBurstLimit
                    }
                },
                accessLogDestination: new apigw.LogGroupLogDestination(logGroup),
                accessLogFormat: apigw.AccessLogFormat.clf()
            }
        });

        const validator = this.Api.addRequestValidator('Validator', {
            requestValidatorName: 'default-validator',
            validateRequestBody: true
        });

        this.addIntegration({
            action: 'PutRecord',
            path: 'record',
            model: this.getPutRecordModel(),
            requestTemplate: this.getPutRecordRequestTemplate(props.stream.streamName),
            role: apiGatewayRole,
            validator
        });

        this.addIntegration({
            action: 'PutRecords',
            path: 'records',
            model: this.getPutRecordsModel(),
            requestTemplate: this.getPutRecordsRequestTemplate(props.stream.streamName),
            role: apiGatewayRole,
            validator
        });

        this.addCfnNagSuppressions();
    }

    private getPutRecordRequestTemplate(streamName: string): string {
        return `{ "StreamName": "${streamName}", "Data": "$util.base64Encode($input.json('$.data'))", "PartitionKey": "$input.path('$.partitionKey')" }`;
    }

    private getPutRecordModel(): apigw.IModel {
        return this.Api.addModel('PutRecordModel', {
            contentType: 'application/json',
            modelName: 'PutRecordModel',
            description: 'PutRecord proxy single-record payload',
            schema: {
                schema: apigw.JsonSchemaVersion.DRAFT4,
                title: 'PutRecord proxy single-record payload',
                type: apigw.JsonSchemaType.OBJECT,
                required: ['data', 'partitionKey'],
                properties: {
                    data: { type: apigw.JsonSchemaType.STRING },
                    partitionKey: { type: apigw.JsonSchemaType.STRING }
                }
            }
        });
    }

    private getPutRecordsRequestTemplate(streamName: string): string {
        return `{ "StreamName": "${streamName}", "Records": [ #foreach($elem in $input.path('$.records')) { "Data": "$util.base64Encode($elem.data)", "PartitionKey": "$elem.partitionKey"}#if($foreach.hasNext),#end #end ] }`;
    }

    private getPutRecordsModel(): apigw.IModel {
        return this.Api.addModel('PutRecordsModel', {
            contentType: 'application/json',
            modelName: 'PutRecordsModel',
            description: 'PutRecords proxy payload data',
            schema: {
                schema: apigw.JsonSchemaVersion.DRAFT4,
                title: 'PutRecords proxy payload data',
                type: apigw.JsonSchemaType.OBJECT,
                required: ['records'],
                properties: {
                    records: {
                        type: apigw.JsonSchemaType.ARRAY,
                        items: {
                            type: apigw.JsonSchemaType.OBJECT,
                            required: ['data', 'partitionKey'],
                            properties: {
                                data: { type: apigw.JsonSchemaType.STRING },
                                partitionKey: { type: apigw.JsonSchemaType.STRING }
                            }
                        }
                    }
                }
            }
        });
    }

    private addIntegration(props: IntegrationProps)  {
        const integration = new apigw.AwsIntegration({
            service: 'kinesis',
            integrationHttpMethod: 'POST',
            action: props.action,
            options: {
                credentialsRole: props.role,
                passthroughBehavior: apigw.PassthroughBehavior.NEVER,
                requestParameters: {
                    'integration.request.header.Content-Type': "'x-amz-json-1.1'"
                },
                requestTemplates: {
                    'application/json': props.requestTemplate
                },
                integrationResponses: [
                    {
                        statusCode: '200'
                    },
                    {
                        statusCode: '500',
                        selectionPattern: '5\\d{2}'
                    }
                ]
            }
        });

        const resource = this.Api.root.addResource(props.path, {
            defaultCorsPreflightOptions: {
                allowOrigins: apigw.Cors.ALL_ORIGINS,
                allowMethods: ['PUT']
            }
        });

        resource.addMethod('PUT', integration, {
            authorizationType: apigw.AuthorizationType.IAM,
            requestModels: {
                'application/json': props.model
            },
            requestValidator: props.validator,
            methodResponses: [
                {
                    statusCode: '200',
                    responseParameters: { 'method.response.header.Content-Type': true }
                },
                {
                    statusCode: '500',
                    responseParameters: { 'method.response.header.Content-Type': true }
                }
            ]
        });
    }

    private addCfnNagSuppressions() {
        const cfnDeployment = this.Api.latestDeployment?.node.defaultChild as apigw.CfnDeployment;
        cfnDeployment.cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W68',
                    reason: 'Default usage plan can be used for this API'
                }]
            }
        };

        const cfnStage = this.Api.deploymentStage.node.defaultChild as apigw.CfnStage;
        cfnStage.cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W64',
                    reason: 'Default usage plan can be used for this API'
                }]
            }
        };
    }
}
