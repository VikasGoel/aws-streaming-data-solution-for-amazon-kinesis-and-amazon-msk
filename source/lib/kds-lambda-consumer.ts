/*********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as kinesis from '@aws-cdk/aws-kinesis';
import * as sqs from '@aws-cdk/aws-sqs';
import * as cw from '@aws-cdk/aws-cloudwatch';

import { KinesisEventSource, SqsDlq } from '@aws-cdk/aws-lambda-event-sources';
import { ExecutionRole } from './lambda-role-cloudwatch';

export interface LambdaConsumerProps {
    readonly stream: kinesis.IStream;
    readonly batchSize: number;
    readonly retryAttempts: number;
    readonly parallelizationFactor: number;
    readonly timeout: cdk.Duration;
}

export class LambdaConsumer extends cdk.Construct {
    public readonly Function: lambda.IFunction;

    private MIN_TIMEOUT_SECONDS: number = 1;
    private MAX_TIMEOUT_SECONDS: number = 900;

    constructor(scope: cdk.Construct, id: string, props: LambdaConsumerProps) {
        super(scope, id);

        const timeoutSeconds = props.timeout.toSeconds();
        if (timeoutSeconds < this.MIN_TIMEOUT_SECONDS || timeoutSeconds > this.MAX_TIMEOUT_SECONDS) {
            throw new Error(`timeout must be a value between ${this.MIN_TIMEOUT_SECONDS} and ${this.MAX_TIMEOUT_SECONDS} seconds`);
        }

        const dlq = this.createDeadLetterQueue();
        const executionRole = new ExecutionRole(this, 'Role');

        this.Function = new lambda.Function(this, 'Consumer', {
            runtime: lambda.Runtime.NODEJS_12_X,
            handler: 'index.handler',
            role: executionRole.Role,
            code: lambda.Code.fromAsset('lambda/kds-lambda-consumer'),
            timeout: props.timeout
        });

        this.Function.addEventSource(new KinesisEventSource(props.stream, {
            startingPosition: lambda.StartingPosition.LATEST,
            batchSize: props.batchSize,
            retryAttempts: props.retryAttempts,
            parallelizationFactor: props.parallelizationFactor,
            bisectBatchOnError: true,
            onFailure: new SqsDlq(dlq)
        }));
    }

    private createDeadLetterQueue(): sqs.Queue {
        const dlq = new sqs.Queue(this, 'DLQ', {
            encryption: sqs.QueueEncryption.KMS_MANAGED,
            dataKeyReuse: cdk.Duration.minutes(30)
        });

        new cw.Alarm(this, 'DLQDepthAlarm', {
            metric: new cw.Metric({
                namespace: 'AWS/SQS',
                metricName: 'ApproximateNumberOfMessagesVisible',
                dimensions: { 'QueueName': dlq.queueName },
                statistic: 'Sum',
                period: cdk.Duration.minutes(5)
            }),
            threshold: 1,
            comparisonOperator: cw.ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
            evaluationPeriods: 1,
            treatMissingData: cw.TreatMissingData.NOT_BREACHING
        });

        return dlq;
    }
}