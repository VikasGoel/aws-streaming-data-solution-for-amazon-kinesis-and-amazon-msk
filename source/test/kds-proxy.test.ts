/*********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from '@aws-cdk/core';
import * as logs from '@aws-cdk/aws-logs';
import * as kinesis from '@aws-cdk/aws-kinesis';
import { expect as expectCDK, haveResource, haveResourceLike, ResourcePart, SynthUtils } from '@aws-cdk/assert';

import { ProxyApi } from '../lib/kds-apigw-proxy';

describe('successful scenarios', () => {
    let stack: cdk.Stack;

    beforeAll(() => {
        const app = new cdk.App();
        stack = new cdk.Stack(app, 'TestStack');

        new ProxyApi(stack, 'TestApi', {
            stream: new kinesis.Stream(stack, 'TestStream'),
            accessLogsRetention: logs.RetentionDays.ONE_WEEK,
            throttlingRateLimit: 100,
            throttlingBurstLimit: 50
        });
    });

    test('creates a proxy API for Kinesis', () => {
        expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    });

    test('configures logging', () => {
        expectCDK(stack).to(haveResource('AWS::Logs::LogGroup', {
            RetentionInDays: 7
        }));

        expectCDK(stack).to(haveResource('AWS::ApiGateway::Account'));
    });

    test('creates an IAM role for API Gateway', () => {
        expectCDK(stack).to(haveResource('AWS::IAM::Role', {
            AssumeRolePolicyDocument: {
                Statement: [{
                    Action: 'sts:AssumeRole',
                    Effect: 'Allow',
                    Principal: {
                        Service: 'apigateway.amazonaws.com'
                    }
                }],
                Version: '2012-10-17'
            }
        }));

        expectCDK(stack).to(haveResourceLike('AWS::IAM::Policy', {
            PolicyDocument: {
                Statement: [{
                    Action: ['kinesis:ListShards', 'kinesis:PutRecord', 'kinesis:PutRecords'],
                    Effect: 'Allow'
                }]
            }
        }));
    });

    test('creates deployment related resources', () => {
        expectCDK(stack).to(haveResource('AWS::ApiGateway::Deployment'));
        expectCDK(stack).to(haveResource('AWS::ApiGateway::Stage', {
            MethodSettings: [{
                HttpMethod: '*',
                ResourcePath: '/*',
                ThrottlingRateLimit: 100,
                ThrottlingBurstLimit: 50
            }],
            StageName: 'prod'
        }));

        expectCDK(stack).to(haveResource('AWS::ApiGateway::RequestValidator', {
            ValidateRequestBody: true
        }));

        expectCDK(stack).to(haveResource('AWS::ApiGateway::Model', {
            Name: 'PutRecordModel'
        }));

        expectCDK(stack).to(haveResource('AWS::ApiGateway::Model', {
            Name: 'PutRecordsModel'
        }));
    });

    test('adds cfn_nag suppressions', () => {
        expectCDK(stack).to(haveResource('AWS::ApiGateway::Deployment', {
            Metadata: {
                cfn_nag: {
                    rules_to_suppress: [{
                        id: 'W68',
                        reason: 'Default usage plan can be used for this API'
                    }]
                }
            }
        }, ResourcePart.CompleteDefinition));

        expectCDK(stack).to(haveResource('AWS::ApiGateway::Stage', {
            Metadata: {
                cfn_nag: {
                    rules_to_suppress: [{
                        id: 'W64',
                        reason: 'Default usage plan can be used for this API'
                    }]
                }
            }
        }, ResourcePart.CompleteDefinition));
    });
});

describe('validation tests', () => {
    let stack: cdk.Stack;
    let fakeStream: kinesis.IStream;

    beforeEach(() => {
        const app = new cdk.App();
        stack = new cdk.Stack(app, 'TestStack');
        fakeStream = new kinesis.Stream(stack, 'TestStream');
    });

    test.each([0, -1])('rate limit must be positive', (invalidRateLimit) => {
        expect(() => new ProxyApi(stack, 'TestApi', {
            stream: fakeStream,
            accessLogsRetention: logs.RetentionDays.ONE_WEEK,
            throttlingRateLimit: invalidRateLimit,
            throttlingBurstLimit: 50
        })).toThrowError(/throttlingRateLimit must be a positive number/);
    });

    test('burst limit must not be negative', () => {
        expect(() => new ProxyApi(stack, 'TestApi', {
            stream: fakeStream,
            accessLogsRetention: logs.RetentionDays.ONE_WEEK,
            throttlingRateLimit: 100,
            throttlingBurstLimit: -1
        })).toThrowError(/throttlingBurstLimit must be a non-negative number/);
    });
});
