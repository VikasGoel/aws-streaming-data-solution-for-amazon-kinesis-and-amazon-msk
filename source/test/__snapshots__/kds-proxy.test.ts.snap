// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`successful scenarios creates a proxy API for Kinesis 1`] = `
Object {
  "Conditions": Object {
    "AwsCdkKinesisEncryptedStreamsUnsupportedRegions": Object {
      "Fn::Or": Array [
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-north-1",
          ],
        },
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-northwest-1",
          ],
        },
      ],
    },
  },
  "Outputs": Object {
    "TestApiEndpointC789B270": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "TestApiC9C18F5F",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "TestApiDeploymentStageprodF668568D",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Resources": Object {
    "TestApiAccessLogsC7BA5491": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "RetentionInDays": 7,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "TestApiAccount4BDA02C2": Object {
      "DependsOn": Array [
        "TestApiC9C18F5F",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "TestApiCloudWatchRoleA3BEC446",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "TestApiApiGatewayRole5819F962": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TestApiApiGatewayRoleDefaultPolicy76B65131": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kinesis:ListShards",
                "kinesis:PutRecord",
                "kinesis:PutRecords",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TestStreamE6F40222",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestApiApiGatewayRoleDefaultPolicy76B65131",
        "Roles": Array [
          Object {
            "Ref": "TestApiApiGatewayRole5819F962",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestApiC9C18F5F": Object {
      "Properties": Object {
        "Name": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "AWS::StackName",
              },
              "-kinesis-proxy",
            ],
          ],
        },
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "TestApiCloudWatchRoleA3BEC446": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestApiDeploymentBC0C43ADd69b57fd15a7b4bd0bce05ee008c060a": Object {
      "DependsOn": Array [
        "TestApirecordOPTIONS16FD6C23",
        "TestApirecordPUT7D5C36EC",
        "TestApirecord6E9B2923",
        "TestApirecordsOPTIONS50080610",
        "TestApirecordsPUTEB989000",
        "TestApirecordsCA61DDD3",
        "TestApiPutRecordModel948379D1",
        "TestApiPutRecordsModelD298C048",
        "TestApiValidator606D2DAC",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W68",
              "reason": "Default usage plan can be used for this API",
            },
          ],
        },
      },
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "TestApiC9C18F5F",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "TestApiDeploymentStageprodF668568D": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W64",
              "reason": "Default usage plan can be used for this API",
            },
          ],
        },
      },
      "Properties": Object {
        "AccessLogSetting": Object {
          "DestinationArn": Object {
            "Fn::GetAtt": Array [
              "TestApiAccessLogsC7BA5491",
              "Arn",
            ],
          },
          "Format": "$context.identity.sourceIp $context.identity.caller $context.identity.user [$context.requestTime] \\"$context.httpMethod $context.resourcePath $context.protocol\\" $context.status $context.responseLength $context.requestId",
        },
        "DeploymentId": Object {
          "Ref": "TestApiDeploymentBC0C43ADd69b57fd15a7b4bd0bce05ee008c060a",
        },
        "MethodSettings": Array [
          Object {
            "HttpMethod": "*",
            "ResourcePath": "/*",
            "ThrottlingBurstLimit": 50,
            "ThrottlingRateLimit": 100,
          },
        ],
        "RestApiId": Object {
          "Ref": "TestApiC9C18F5F",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "TestApiPutRecordModel948379D1": Object {
      "Properties": Object {
        "ContentType": "application/json",
        "Description": "PutRecord proxy single-record payload",
        "Name": "PutRecordModel",
        "RestApiId": Object {
          "Ref": "TestApiC9C18F5F",
        },
        "Schema": Object {
          "$schema": "http://json-schema.org/draft-04/schema#",
          "properties": Object {
            "data": Object {
              "type": "string",
            },
            "partitionKey": Object {
              "type": "string",
            },
          },
          "required": Array [
            "data",
            "partitionKey",
          ],
          "title": "PutRecord proxy single-record payload",
          "type": "object",
        },
      },
      "Type": "AWS::ApiGateway::Model",
    },
    "TestApiPutRecordsModelD298C048": Object {
      "Properties": Object {
        "ContentType": "application/json",
        "Description": "PutRecords proxy payload data",
        "Name": "PutRecordsModel",
        "RestApiId": Object {
          "Ref": "TestApiC9C18F5F",
        },
        "Schema": Object {
          "$schema": "http://json-schema.org/draft-04/schema#",
          "properties": Object {
            "records": Object {
              "items": Object {
                "properties": Object {
                  "data": Object {
                    "type": "string",
                  },
                  "partitionKey": Object {
                    "type": "string",
                  },
                },
                "required": Array [
                  "data",
                  "partitionKey",
                ],
                "type": "object",
              },
              "type": "array",
            },
          },
          "required": Array [
            "records",
          ],
          "title": "PutRecords proxy payload data",
          "type": "object",
        },
      },
      "Type": "AWS::ApiGateway::Model",
    },
    "TestApiValidator606D2DAC": Object {
      "Properties": Object {
        "Name": "default-validator",
        "RestApiId": Object {
          "Ref": "TestApiC9C18F5F",
        },
        "ValidateRequestBody": true,
      },
      "Type": "AWS::ApiGateway::RequestValidator",
    },
    "TestApirecord6E9B2923": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "TestApiC9C18F5F",
            "RootResourceId",
          ],
        },
        "PathPart": "record",
        "RestApiId": Object {
          "Ref": "TestApiC9C18F5F",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "TestApirecordOPTIONS16FD6C23": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "OPTIONS",
        "Integration": Object {
          "IntegrationResponses": Array [
            Object {
              "ResponseParameters": Object {
                "method.response.header.Access-Control-Allow-Headers": "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token,X-Amz-User-Agent'",
                "method.response.header.Access-Control-Allow-Methods": "'PUT'",
                "method.response.header.Access-Control-Allow-Origin": "'*'",
              },
              "StatusCode": "204",
            },
          ],
          "RequestTemplates": Object {
            "application/json": "{ statusCode: 200 }",
          },
          "Type": "MOCK",
        },
        "MethodResponses": Array [
          Object {
            "ResponseParameters": Object {
              "method.response.header.Access-Control-Allow-Headers": true,
              "method.response.header.Access-Control-Allow-Methods": true,
              "method.response.header.Access-Control-Allow-Origin": true,
            },
            "StatusCode": "204",
          },
        ],
        "ResourceId": Object {
          "Ref": "TestApirecord6E9B2923",
        },
        "RestApiId": Object {
          "Ref": "TestApiC9C18F5F",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "TestApirecordPUT7D5C36EC": Object {
      "Properties": Object {
        "AuthorizationType": "AWS_IAM",
        "HttpMethod": "PUT",
        "Integration": Object {
          "Credentials": Object {
            "Fn::GetAtt": Array [
              "TestApiApiGatewayRole5819F962",
              "Arn",
            ],
          },
          "IntegrationHttpMethod": "POST",
          "IntegrationResponses": Array [
            Object {
              "StatusCode": "200",
            },
            Object {
              "SelectionPattern": "5\\\\d{2}",
              "StatusCode": "500",
            },
          ],
          "PassthroughBehavior": "NEVER",
          "RequestParameters": Object {
            "integration.request.header.Content-Type": "'x-amz-json-1.1'",
          },
          "RequestTemplates": Object {
            "application/json": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "{ \\"StreamName\\": \\"",
                  Object {
                    "Ref": "TestStreamE6F40222",
                  },
                  "\\", \\"Data\\": \\"$util.base64Encode($input.json('$.data'))\\", \\"PartitionKey\\": \\"$input.path('$.partitionKey')\\" }",
                ],
              ],
            },
          },
          "Type": "AWS",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":kinesis:action/PutRecord",
              ],
            ],
          },
        },
        "MethodResponses": Array [
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "200",
          },
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "500",
          },
        ],
        "RequestModels": Object {
          "application/json": Object {
            "Ref": "TestApiPutRecordModel948379D1",
          },
        },
        "RequestValidatorId": Object {
          "Ref": "TestApiValidator606D2DAC",
        },
        "ResourceId": Object {
          "Ref": "TestApirecord6E9B2923",
        },
        "RestApiId": Object {
          "Ref": "TestApiC9C18F5F",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "TestApirecordsCA61DDD3": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "TestApiC9C18F5F",
            "RootResourceId",
          ],
        },
        "PathPart": "records",
        "RestApiId": Object {
          "Ref": "TestApiC9C18F5F",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "TestApirecordsOPTIONS50080610": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "OPTIONS",
        "Integration": Object {
          "IntegrationResponses": Array [
            Object {
              "ResponseParameters": Object {
                "method.response.header.Access-Control-Allow-Headers": "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token,X-Amz-User-Agent'",
                "method.response.header.Access-Control-Allow-Methods": "'PUT'",
                "method.response.header.Access-Control-Allow-Origin": "'*'",
              },
              "StatusCode": "204",
            },
          ],
          "RequestTemplates": Object {
            "application/json": "{ statusCode: 200 }",
          },
          "Type": "MOCK",
        },
        "MethodResponses": Array [
          Object {
            "ResponseParameters": Object {
              "method.response.header.Access-Control-Allow-Headers": true,
              "method.response.header.Access-Control-Allow-Methods": true,
              "method.response.header.Access-Control-Allow-Origin": true,
            },
            "StatusCode": "204",
          },
        ],
        "ResourceId": Object {
          "Ref": "TestApirecordsCA61DDD3",
        },
        "RestApiId": Object {
          "Ref": "TestApiC9C18F5F",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "TestApirecordsPUTEB989000": Object {
      "Properties": Object {
        "AuthorizationType": "AWS_IAM",
        "HttpMethod": "PUT",
        "Integration": Object {
          "Credentials": Object {
            "Fn::GetAtt": Array [
              "TestApiApiGatewayRole5819F962",
              "Arn",
            ],
          },
          "IntegrationHttpMethod": "POST",
          "IntegrationResponses": Array [
            Object {
              "StatusCode": "200",
            },
            Object {
              "SelectionPattern": "5\\\\d{2}",
              "StatusCode": "500",
            },
          ],
          "PassthroughBehavior": "NEVER",
          "RequestParameters": Object {
            "integration.request.header.Content-Type": "'x-amz-json-1.1'",
          },
          "RequestTemplates": Object {
            "application/json": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "{ \\"StreamName\\": \\"",
                  Object {
                    "Ref": "TestStreamE6F40222",
                  },
                  "\\", \\"Records\\": [ #foreach($elem in $input.path('$.records')) { \\"Data\\": \\"$util.base64Encode($elem.data)\\", \\"PartitionKey\\": \\"$elem.partitionKey\\"}#if($foreach.hasNext),#end #end ] }",
                ],
              ],
            },
          },
          "Type": "AWS",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":kinesis:action/PutRecords",
              ],
            ],
          },
        },
        "MethodResponses": Array [
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "200",
          },
          Object {
            "ResponseParameters": Object {
              "method.response.header.Content-Type": true,
            },
            "StatusCode": "500",
          },
        ],
        "RequestModels": Object {
          "application/json": Object {
            "Ref": "TestApiPutRecordsModelD298C048",
          },
        },
        "RequestValidatorId": Object {
          "Ref": "TestApiValidator606D2DAC",
        },
        "ResourceId": Object {
          "Ref": "TestApirecordsCA61DDD3",
        },
        "RestApiId": Object {
          "Ref": "TestApiC9C18F5F",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "TestStreamE6F40222": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "Fn::If": Array [
            "AwsCdkKinesisEncryptedStreamsUnsupportedRegions",
            Object {
              "Ref": "AWS::NoValue",
            },
            Object {
              "EncryptionType": "KMS",
              "KeyId": "alias/aws/kinesis",
            },
          ],
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
  },
}
`;
